spec.bs

```diff
@@ -6,57 +6,417 @@ Status: CG-DRAFT
+        1. Let |index_promise| be a new promise.
+        1. Let |worklet_queue| be the <a href="https://html.spec.whatwg.org/multipage/infrastructure.html#parallel-queue">parallel queue</a> for the context's associated {{SharedStorageWorklet}} interface, if it exists.
+        1. If |worklet_queue| does not exist, create it.
+        1. <a href="https://html.spec.whatwg.org/multipage/infrastructure.html#enqueue-the-following-steps">Enqueue the following steps</a> on |worklet_queue|            1. Let |index| be 0.
+            1. If module script has been loaded for this worklet:
+                1. Let |operation_map| be the operation map in the associated {{SharedStorageWorkletGlobalScope}}.
+                1. Let |operation| be the operation registered to |name| in |operation_map|, if |operation| exists.
+                1. If |operation| exists:
+                    1. If |options| contains |data|, let |operation_result| be the result of invoking |operation| with arguments |url_list| and |data|.
+                    1. Otherwise, let |operation_result| be the result of invoking |operation| with argument |url_list|.
+                    1. If there are no errors:
+                        1. Set |index| to the result of casting |operation_result| to an {{unsigned long}}. If this throws an exception, catch it and set |index| back to 0.
+                        1. If |index| > |url_list|.[=list/size=], then set |index| back to 0.
+            1. Resolve |index_promise| with |index|.
+        1. When |index_promise| is resolved, let |result_index| be its numerical value.
+        1. Let |remaining_budget| be the numerical result from {{SharedStorage/remainingBudget()}}.
+        1. Let |current_debit| be log_2(|url_list|.[=list/size=]).
+        1. If |current_debit| > |remaining_budget|, set |result_index| to 0.
+        1. Store |current_debit| so that it can be withdrawn from the database in the event of a frame created from |fenced_frame_config| triggering a top-frame navigation.
+        1. Internally map |pending_url| to |url_list|[|result_index|].
+  </div>
+
+  ## Setter/Deleter Methods ## {#setter}
+
+  <div algorithm>
+    The <dfn method for="SharedStorage">set(|key|, |value|, |options|)</dfn> method steps are:
+
+    1. Let |promise| be a new promise.
+    1. <a href="https://html.spec.whatwg.org/multipage/infrastructure.html#enqueue-the-following-steps">Enqueue the following steps</a>:
+        1. If |key| is invalid, reject |promise| with a "{{DataError}}" {{DOMException}}.
+        1. If |value| is invalid, reject |promise| with a "{{DataError}}" {{DOMException}}.
+        1. If |options| is invalid, reject |promise| with a "{{DataError}}" {{DOMException}}.
+        1. If |options| is empty, create |options| with "|ignoreIfPresent|" set to false.
+        1. Perform a database query to determine whether an |entry| for |key| already exists in the database for this context's origin.
```

Comment:
More rigorously define the database types and accessors. Ideally we probably should follow the terms similar to indexed db specification:
https://www.w3.org/TR/IndexedDB/#object-store
https://www.w3.org/TR/IndexedDB/#object-store-storage-operation

But as a start / for simplicity, we can model the storage as a key-value type. And define it as an user agent associated state. And perhaps add a note that "the browser should use a database backend for accessing the storage."

Each [=user agent=] has an associated <dfn for="user agent">shared storage area</dfn>, it is a map.

(When accessing it)
If user agent's [=user agent/shared storage area=] does not contain |origin|:
	- Set user agent's [=user agent/shared storage area=][|origin|] to an empty map.

Let XXX be agent's [=user agent/shared storage area=][origin],
(access the value for a key, etc…)

Reply:
You should probably hook this into the storage spec data structures:

https://storage.spec.whatwg.org/#model

I believe the storage bucket contains a map from storage type strings (like "indexeddb") to storage bottles.

So you would monkey patch in a new registered storage type for "sharedstorage" or something, then put your map of data inside the resulting storage bottle you get.